File: src/easy/IsPalindrome.java
package easy;

import common.ListNode;

public class IsPalindrome {

    public boolean isPalindrome(int x) {
        if (x < 0 || (x != 0 && x % 10 == 0)) {
            return false;
        }

        int rev = 0;
        while (x > rev) {
            // rev * 10 and add the tail of x
            rev = rev * 10 + x % 10;
            x /= 10;
        }
        // even length x = rev, odd length x = rev / 10
        return (x == rev) || (x == rev / 10);
    }

    public boolean isPalindrome(ListNode head) {
        assert head != null : "asdf";
        if (head == null || head.next == null) {
            return true;
        }
        ListNode mid = findMid(head);
        ListNode right = mid.next;
        mid.next = null;
        ListNode reversedRight = reverse(right);
        return compare(head, reversedRight);
    }

    public boolean isPalindrome(String s) {
        if (s == null || s.length() == 0) {
            return true;
        }
        char[] chars = s.toLowerCase().toCharArray();
        int l = 0;
        int r = s.length() - 1;
        while (l < r) {
            while (l < r && !Character.isLetterOrDigit(chars[l])) {
                l++;
            }
            while (l < r && !Character.isLetterOrDigit(chars[r])) {
                r++;
            }
            if (l < r && chars[l] != chars[r]) {
                return false;
            }
        }
        return true;
    }

    private boolean compare(ListNode head, ListNode reversedRight) {
        while (head != null && reversedRight != null) {
            if (head.val != reversedRight.val) {
                return false;
            }
            head = head.next;
            reversedRight = reversedRight.next;
        }
        return true;
    }

    private ListNode reverse(ListNode node) {
        ListNode left = null;
        ListNode cur = node;
        while (cur != null) {
            ListNode right = cur.next;
            cur.next = left;
            left = cur;
            cur = right;
        }
        return left;
    }

    private ListNode findMid(ListNode head) {
        ListNode fast = head.next;
        ListNode slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }


}

File: csharp/LeetCode/Easy/IsPalindrome.cs
namespace LeetCode.Easy.IsPalindrome
{
    public class Solution
    {
        public bool IsPalindrome(string s)
        {
            var l = 0;
            var r = s.Length - 1;

            while (l < r)
            {
                if (!char.IsLetterOrDigit(s[l]))
                {
                    l++;
                }
                else if (!char.IsLetterOrDigit(s[r]))
                {
                    r--;
                }
                else
                {
                    if (char.ToLower(s[l]) != char.ToLower(s[r]))
                    {
                        return false;
                    }
                    l++;
                    r--;
                }
            }
            return true;
        }

        public bool IsPalindrome(int n)
        {
            if (n < 0)
            {
                return false;
            }
            var x = n;
            var y = 0;
            while (x > 0)
            {
                var r = x % 10;
                y = y * 10 + r;
                x /= 10;
            }
            return n == y;
        }
    }
}
